_posts/2020-04-18-methods-arguments-ruby-bart3.markdown

```diff
@@ -0,0 +1,257 @@
+
+foo('arg1') do |message|
+  puts message
+end
+# => arg1
+```
+
+## Array Deconstruction
+
+Sometimes we may have a method that accepts an array as one of the arguments. Usually we just use a normal argument and extract the values inside the method:
+
+```ruby
+def foo(my_arr)
+  el1, el2 = my_arr
+  # ...
+end
+
+foo([1,2])
+```
+
+But we can do that when defining parameters so we don't need that line and we get the benefit of being able to use the array elements when defining the default value of other arguments:
+
+```ruby
+def foo((el1, el2), arg2 = el1 * 5)
+  puts el1
+  puts el2
+  puts arg2
+end
+
+foo([1, 2])
+# => 1
+#    2
+#    5
+
+foo([1, 2], 'something')
+# => 1
+#    2
+#    something
+```
+
+> If our array includes more elements than the expected, the rest are discarded!
+
+## Ignoring Arguments
+
+If you want to accept any number of arguments but not use them (imagine you are implementing a replacement of a library you are using and you have to respect the method calls but don't really need the arguments) you can use the `*` and `**` operators with no name for the parameter:
+
+```ruby
+def foo(*, **)
+  puts 'I ignore everything'
+end
+
+foo(1, 2, 'three', more_things: 'ignored')
+# => I ignore everything
+```
+
+## Arguments Delegation
+
+If you are writing a wrapper around another method, you usually need to accept the same arguments, do something, and then call the original method. If we need to use exactly the same arguments and we don't need them for our extra code, we can use a special argument for delegation introduced in Ruby 2.7 (the `...` operator):
+
+```ruby
+def foo(...)
+  bar(...)
+end
+
+def bar(arg1, arg2)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+end
+
+foo(1, 2)
+# => arg1 is: 1
+#    arg2 is: 2
+```
+
+## Partially Applied Methods
+
+We may have a process that does multiple calls to one method using mostly the same arguments and only changes some of them. To refactor those calls we can use the concept of function (or method) generators. The idea is to get our method, apply only some arguments, and then have a new method that requires only the rest of the arguments that were not applied yet.
+
+Let's say we have a generic method to multiply two numbers:
+
+```ruby
+def multi(num1, num2)
+  puts num1 * num2
+end
+```
+
+Now let's imagine we want to have a method that multiplies by 3 because we are using that a lot. We can reuse that original method and only apply the number 3 using the `curry` method on the method object:
+
+```ruby
+multi3 = method(:multi).curry.call(3)
+#        ^ we use method(:mutli) to get the method object
+#          instead of calling the method
+```
+
+Now we can use that new method (it's actually a Proc) to have different numbers multipied by 3:
+
+```ruby
+multi3.call(2) # => 6
+multi3.call(5) # => 15
+
+# .() is a shorthand for .call()
+multi3.(10) # => 30
+```
+
+## Case Study 1: link_to
+
+Let's analyze Rails' `link_to` method.
```

Comment:
@arielj It would be nice to put the link to the github file where you extracted this